### PR TITLE
Feature: Add the ability to parse text into spans

### PR DESCRIPTION
**Pull request by KennedyTedesco**
Following: https://github.com/php-tui/php-tui/pull/136

With this new implementation, users now have two options for formatting their texts:

1) Using style methods
2) Parsing directly

So, this style:

image

Can be achieved by doing either:

```php
Line::fromSpans([
    Span::fromString('Hello ')->black()->onBlue()->bold()->italic(),
    Span::fromString('Wor')->white()->onRed()->crossedout(),
    Span::fromString('ld')->black()->onBlue()->bold()->italic(),
    Span::fromString(' PHP'),
]),
```

or:

```php
Line::parse('<fg=black;bg=blue;options=bold,italic>Hello <fg=white;bg=red;options=crossedout>Wor</>ld</> PHP')
```

Other examples:

```php
// Style methods
$title = Title::fromLine(Line::fromSpans([
    Span::fromString('Titles')->red()->onWhite(),
]));

// Parsing instead:
$title = Title::parse('<fg=red;bg=white>Titles</>');
```

### CODE REVIEW

File: src/Model/Parseable.php [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+declare(strict_types=1);
+
+namespace PhpTui\Tui\Model;
+
+interface Parseable
+{
+    public static function parse(string $string): self;
```
Comment (dantleech) [inline src/Model/Parseable.php:9]: This interface is a bit anaemic (it's about Style but the naming implies anything or everything) and it's role is more of a Parser and not something that can be parsed? - maybe rename to StyleParser

**Comment by dantleech** [comment]
Nice work! Looks good from a quick once over ,will have a better look later.

File: src/Model/Parseable.php (hunk 1 above)
Comment (KennedyTedesco) [reply src/Model/Parseable.php:9]: I'm out of the computer right now, but, yeah. To be honest, I don't even know if it makes sense to have this interface at all. Thinking about dropping it for now... Unless you want me to keep it or change it. Can you please review this later (where it is used etc) and give me some feedback about this?

File: example/demo/src/Page/BlocksPage.php [hunk 2]
```diff
@@ -80,10 +78,7 @@ public function lorem(): ParagraphWidget
         $text = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.';
 
         return ParagraphWidget::fromText(
-            Text::styled(
-                $text,
-                Style::default()->fg(AnsiColor::DarkGray)
-            )
+            Text::parse("<fg=darkgray>{$text}</>")
```
Comment (dantleech) [inline example/demo/src/Page/BlocksPage.php:81]: minor: use `sprintf` instead

File: src/Model/Widget/Line.php [hunk 3]
```diff
@@ -58,6 +59,13 @@ public static function fromString(string $string): self
         ], null);
     }
 
+    public static function parse(string $string): self
+    {
+        return self::fromSpans(
+            SpanParser::new()->parse($string)
```
Comment (dantleech) [inline src/Model/Widget/Line.php:65]: not related to this PR but I want to make all `fromFoos(` variadics, so should be `fromSpans(...SpanParser::new()->parse($string))`

File: src/Model/Widget/SpanParser.php [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+use PhpTui\Tui\Model\Modifier;
+use PhpTui\Tui\Model\Style;
+
+final class SpanParser
+{
+    private const OPEN_TAG_REGEX = '[a-z](?:[^\\\\<>]*+ | \\\\.)*';
+    private const CLOSE_TAG_REGEX = '[a-z][^<>]*+';
+
+    /** @var array<Style> */
+    private array $styleStack = [];
```
Comment (dantleech) [inline src/Model/Widget/SpanParser.php:19]: maybe avoid state and pass `&$styleStack` in the function calls?

File: src/Model/Widget/SpanParser.php [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+use PhpTui\Tui\Model\Color;
+use PhpTui\Tui\Model\Color\AnsiColor;
+use PhpTui\Tui\Model\Color\RgbColor;
+use PhpTui\Tui\Model\Modifier;
+use PhpTui\Tui\Model\Style;
+
+final class SpanParser
+{
+    private const OPEN_TAG_REGEX = '[a-z](?:[^\\\\<>]*+ | \\\\.)*';
```
Comment (dantleech) [inline src/Model/Widget/SpanParser.php:15]: i'd be lying if i said I understood what this does :eyes:

File: src/Model/Widget/SpanParser.php [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+namespace PhpTui\Tui\Model\Widget;
+
+use PhpTui\Tui\Model\Color;
+use PhpTui\Tui\Model\Color\AnsiColor;
+use PhpTui\Tui\Model\Color\RgbColor;
+use PhpTui\Tui\Model\Modifier;
+use PhpTui\Tui\Model\Style;
+
+final class SpanParser
```
Comment (dantleech) [inline src/Model/Widget/SpanParser.php:13]: let's make this a subset of the Symfony Console component markup, and add a class comment here with a link to: `https://symfony.com/doc/current/console/coloring.html`

File: src/Model/Widget/SpanParser.php [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    {
+        return new self();
+    }
+
+    /**
+     * @return array<Span>
+     */
+    public function parse(string $input): array
+    {
+        $regex = sprintf('#<((%s) | /(%s)?)>#ix', self::OPEN_TAG_REGEX, self::CLOSE_TAG_REGEX);
```
Comment (dantleech) [inline src/Model/Widget/SpanParser.php:31]: in Symfony the delimiter is `;` not space...

File: src/Model/Widget/Text.php [hunk 8]
```diff
@@ -26,6 +27,22 @@ public static function fromString(string $string): self
         }, explode("\n", $string)));
     }
 
+    public static function parse(string $string): self
+    {
+        $placeholder = sprintf('{{new_line_%s}}', microtime(true));
```
Comment (dantleech) [inline src/Model/Widget/Text.php:32]: `uniqid` would be safer than `microtime`

File: src/Model/Widget/SpanParser.php (hunk 7 above)
Comment (KennedyTedesco) [reply src/Model/Widget/SpanParser.php:31]: Yes, I didn't strictly follow them, tbh. Do you prefer `;`?

File: src/Model/Widget/Text.php (hunk 8 above)
Comment (dantleech) [reply src/Model/Widget/Text.php:32]: but can also be a static string: `%%__UNLIKELY_STRING_HERE__%%`

File: src/Model/Widget/SpanParser.php (hunk 4 above)
Comment (KennedyTedesco) [reply src/Model/Widget/SpanParser.php:19]: Sounds reasonable. I'll do it.

File: src/Model/Widget/Line.php (hunk 3 above)
Comment (KennedyTedesco) [reply src/Model/Widget/Line.php:65]: I love this idea!

File: tests/Unit/Model/Widget/SpanParserTest.php [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+use PhpTui\Tui\Model\Color\RgbColor;
+use PhpTui\Tui\Model\Modifier;
+use PhpTui\Tui\Model\Widget\SpanParser;
+use PHPUnit\Framework\TestCase;
+
+class SpanParserTest extends TestCase
+{
+    public function testParseOneTag(): void
+    {
+        $spans = SpanParser::new()->parse('<fg=green bg=blue options=bold,italic>Hello</> World');
```
Comment (dantleech) [inline tests/Unit/Model/Widget/SpanParserTest.php:18]: as mentioned previously, to be compatible with Symfony we need this to be delimited with `;`

File: example/demo/src/Page/BlocksPage.php (hunk 2 above)
Comment (KennedyTedesco) [reply example/demo/src/Page/BlocksPage.php:81]: Yes, it probably exists. I'll see this later.

File: src/Model/Widget/Text.php (hunk 8 above)
Comment (KennedyTedesco) [reply src/Model/Widget/Text.php:32]: Yes, I thought about a static string, but I was concerned about conflicts. Maybe we can use something like: `{{php-tui-new-line-placeholder}}`

File: src/Model/Parseable.php (hunk 1 above)
Comment (dantleech) [reply src/Model/Parseable.php:9]: yeah I don't think it makes sense atm

File: src/Model/Widget/Line.php (hunk 3 above)
Comment (dantleech) [reply src/Model/Widget/Line.php:65]: it should already be the case, but I obviously missed some :smile:

File: tests/Unit/Model/Widget/SpanParserTest.php (hunk 9 above)
Comment (KennedyTedesco) [reply tests/Unit/Model/Widget/SpanParserTest.php:18]: ✅

File: src/Model/Widget/SpanParser.php (hunk 7 above)
Comment (dantleech) [reply src/Model/Widget/SpanParser.php:31]: yes, let's make it the same as Symfony

Comment (KennedyTedesco) [reply src/Model/Widget/SpanParser.php:31]: Deal!

File: src/Model/Widget/Text.php (hunk 8 above)
Comment (dantleech) [reply src/Model/Widget/Text.php:32]: or just `1162128a-269a-4c09-88be-effc70a62a3a` :D

Comment (KennedyTedesco) [reply src/Model/Widget/Text.php:32]: That works haha

File: src/Model/Widget/SpanParser.php [hunk 10]
```diff
@@ -10,6 +10,12 @@
 use PhpTui\Tui\Model\Modifier;
 use PhpTui\Tui\Model\Style;
 
+/**
+ * This class is a subset of the Symfony Console component's OutputFormatter class.
+ * It parses strings containing tags into a list of spans.
```
Comment (dantleech) [inline src/Model/Widget/SpanParser.php:15]: Technically it's not a subset of the class but a subset of the ... markup?

**Review by dantleech — APPROVED** [review]

**Comment by dantleech** [comment]
Nice work!

File: src/Model/Widget/SpanParser.php (hunk 10 above)
Comment (KennedyTedesco) [reply src/Model/Widget/SpanParser.php:15]: Yeah, you're right. 😅

**Pull request merged**